tests/conftest.py

```diff
@@ -1486,6 +1498,124 @@ def collect_db_dump(request, duthosts):
+        port = mg_facts["minigraph_neighbors"][interface]["port"]
+        links[interface] = {
+            "name": neighbor["name"],
+            "host": nbrhosts[neighbor["name"]]["host"],
+            "port": port
+        }
+    find_links(duthost, tbinfo, filter)
+    return links
+
+
+@pytest.fixture(scope="session")
+def ctrl_links(duthost, tbinfo, nbrhosts):
+    assert len(nbrhosts) > 1
+    ctrl_nbr_names = natsort.natsorted(nbrhosts.keys())[:2]
+    logging.info("Controlled links {}".format(ctrl_nbr_names))
+    nbrhosts = {name: nbrhosts[name] for name in ctrl_nbr_names}
+    return find_links_from_nbr(duthost, tbinfo, nbrhosts)
+
+@pytest.fixture(scope="session")
+def macsec_nbrhosts(ctrl_links):
+    return {nbr["name"]: nbr for nbr in ctrl_links.values()}
+
+@pytest.fixture(scope="session")
+def unctrl_links(duthost, tbinfo, nbrhosts, ctrl_links):
+    unctrl_nbr_names = set(nbrhosts.keys())
+    for _, nbr in ctrl_links.items():
+        if nbr["name"] in unctrl_nbr_names:
+            unctrl_nbr_names.remove(nbr["name"])
+    logging.info("Uncontrolled links {}".format(unctrl_nbr_names))
+    nbrhosts = {name: nbrhosts[name] for name in unctrl_nbr_names}
+    return find_links_from_nbr(duthost, tbinfo, nbrhosts)
+
+
+@pytest.fixture(scope="session")
+def downstream_links(duthost, tbinfo, nbrhosts):
+    links = collections.defaultdict(dict)
+
+    def filter(interface, neighbor, mg_facts, tbinfo):
+        if tbinfo["topo"]["type"] == "t0" and "Server" in neighbor["name"]:
+            port = mg_facts["minigraph_neighbors"][interface]["port"]
+            links[interface] = {
+                "name": neighbor["name"],
+                "ptf_port_id": mg_facts["minigraph_ptf_indices"][interface],
+                "port": port
+            }
+    find_links(duthost, tbinfo, filter)
+    return links
+
+
+@pytest.fixture(scope="session")
+def upstream_links(duthost, tbinfo, nbrhosts):
+    links = collections.defaultdict(dict)
+
+    def filter(interface, neighbor, mg_facts, tbinfo):
+        if tbinfo["topo"]["type"] == "t0" and "T1" in neighbor["name"]:
+            for item in mg_facts["minigraph_bgp"]:
+                if item["name"] == neighbor["name"]:
+                    if isinstance(ip_address(item["addr"]), IPv4Address):
+                        local_ipv4_addr = item["addr"]
+                        peer_ipv4_addr = item["peer_addr"]
+                        break
+            port = mg_facts["minigraph_neighbors"][interface]["port"]
+            links[interface] = {
+                "name": neighbor["name"],
+                "ptf_port_id": mg_facts["minigraph_ptf_indices"][interface],
+                "local_ipv4_addr": local_ipv4_addr,
+                "peer_ipv4_addr": peer_ipv4_addr,
+                "port": port,
+                "host": nbrhosts[neighbor["name"]]["host"]
+            }
+    find_links(duthost, tbinfo, filter)
+    return links
+
+
+@pytest.fixture(scope="session")
+def macsec_profile(creds):
```

Comment:
I think it is better to move the dependent fixtures to separate files and import them into this conftest module, including this fixture and some of the dependent fixtures added above.

An example is the dualtor specific fixtures and utilities under `tests/common/dualtor`.

This `conftest.py` is used by all tests, we should try to only add definition of the most frequently used common fixtures into this file.

Reply:
To fix it

Reply:
Fixed